### ISSUE TITLE
Marketplace search: get plugin data by slug

### ISSUE THREAD

**Issue by gcsecsey** [Automattic/wp-calypso#64891]
Depends on #63171 as the data needs to be available.

To be able to fetch data on the plugin pages, we need to add the capability of displaying all plugin details by slug to the marketplace search API.

To do this, we might need to add a new filter query that is run only when a parameter is passed, eg. `marketplace/search/[pluginSlug]`

Or if we create a new endpoint for this:
 - It can be much simpler as the existing one, because it only needs to execute 1 simple query
 - It could reuse/share functions from the existing endpoint

Example ES query to search by slug across both blogs:
```json
GET jetpack-search-21/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "blog_id": ["113771570", "108986944"]
          }
        },
        {
          "term": {
            "slug": {
              "value": "woocommerce"
            }
          }
        }
      ]
    }
  }
}
```

**Comment by cpapazoglou** [comment]
We might also utilise the WC.com API, 12902-gh-Automattic/woocommerce.com if it suits us instead of using ES.

**Comment by gcsecsey** [comment]
Asked whether using ES for this use case is a good idea in pdh6GB-Ax-p2#comment-2677

**Comment by gcsecsey** [comment]
> Asked whether using ES for this use case is a good idea in pdh6GB-Ax-p2#comment-2677

We've got the reply that it's definitely feasible to use ES: pdh6GB-Ax-p2#comment-2679 👍

**Comment by BogdanUngureanu** [comment]
Hey! We currently have an endpoint called `https://public-api.wordpress.com/wpcom/v2/marketplace/products/{productSlug}?_envelope=1` that returns the hardcoded data.

Do we need a new endpoint for this since we already have this endpoint? The way I see it, we should update that endpoint to return the dynamic data. I don't think we need the ES integration, we could just use the shadow site to return the data, since right now that endpoint relies on the product catalog to return the product information.

Imho, this could be handled in Ajax, since we'll be working on removing the coded parts of the catalog anyway. 

@gcsecsey @cpapazoglou what do you think? :)

**Comment by cpapazoglou** [comment]
@BogdanUngureanu the original idea was reusing the same ES endpoint, and thus replacing all Calypso instances where different endpoints are used with this one. But since we will require a second ES endpoint, your suggestion makes sense and should be easier to tackle. I wonder only if using ES would be significant faster?

**Comment by BogdanUngureanu** [comment]
> I wonder only if using ES would be significant faster?

Hmm... Good point! Thinking a bit more, we would need some kind of cache to search by that term value (somehow) since getting by term value would most likely be an expensive operation.

It's something I can look into when I'll rewrite the catalog class.

**Comment by cpapazoglou** [comment]
We will make a research on the indexed data if all fields needed (both for Calypso display, and the product catalog needs) are there and document them here.

**Comment by epeicher** [comment]
We could achieve this functionality only with client-side changes as the `/marketplace/search` endpoints supports some filtering. An example request could be:
```
https://public-api.wordpress.com/rest/v1.3/marketplace/search?http_envelope=1&fields[]=blog_icon_url&fields[]=comment_count&fields[]=plugin.excerpt&fields[]=like_count&fields[]=modified&fields[]=modified_gmt&fields[]=plugin.title&fields[]=author&fields[]=plugin.author&fields[]=author_login&fields[]=blog_id&fields[]=date&fields[]=date_gmt&fields[]=permalink.url.raw&fields[]=post_id&fields[]=post_type&fields[]=slug&fields[]=plugin.tested&fields[]=plugin.stable_tag&fields[]=plugin.active_installs&fields[]=plugin.support_threads&fields[]=plugin.support_threads_resolved&fields[]=plugin.rating&fields[]=plugin.num_ratings&fields[]=plugin.icons&query=&sort=score_default&size=20&lang=en&group_id=wporg&filter[bool][must][][term][slug]=litespeed-cache
```
https://public-api.wordpress.com/rest/v1.3/marketplace/search?http_envelope=1&fields[]=blog_icon_url&fields[]=comment_count&fields[]=plugin.excerpt&fields[]=like_count&fields[]=modified&fields[]=modified_gmt&fields[]=plugin.title&fields[]=author&fields[]=plugin.author&fields[]=author_login&fields[]=blog_id&fields[]=date&fields[]=date_gmt&fields[]=permalink.url.raw&fields[]=post_id&fields[]=post_type&fields[]=slug&fields[]=plugin.tested&fields[]=plugin.stable_tag&fields[]=plugin.active_installs&fields[]=plugin.support_threads&fields[]=plugin.support_threads_resolved&fields[]=plugin.rating&fields[]=plugin.num_ratings&fields[]=plugin.icons&query=&sort=score_default&size=20&lang=en&group_id=wporg&filter[bool][must][][term][slug]=litespeed-cache
Being the filter part the following:
```
&filter[bool][must][][term][slug]=litespeed-cache
```

**Comment by cpapazoglou** [comment]
Wouldn't this return an array of items with just one item? It sounds really good for starters though. Do we already have the fields needed by Calypso?

**Comment by epeicher** [comment]
> Wouldn't this return an array of items with just one item? It sounds really good for starters though. 

Yes, that is correct. In that scenario we can take the first one and assume is the correct one as a starter point.

>Do we already have the fields needed by Calypso?

This is a table of fields that are returned by `/marketplace/products` endpoint nad I have not find in ES:

|Field names returned by `/marketplace/products` |	Field names returned by ES results (wp.org)|
|----|----|
|id	|   :x:   |
|name	|   title.default   |
|slug	|   slug    |
|org_slug	|   :x:   |
|short_description	|   excerpt.en  |
|description	|   description.en  |
|requirements.plugins |	:x:   |
|requirements.required_primary_domain |	:x:   |
|variations.monthly.product_id  |   	:x:   |
|variations.monthly.product_slug  |   	:x:   |
|variations.yearly.product_id |   	:x:   |
|variations.yearly.product_slug |   	:x:   |
|icons.1x|	plugin.icons    |
|icons.2x|	plugin.icons    |
|banners.high|	plugin.banners  |
|banners.low|	plugin.banners  |
|tags	|   taxonomy.plugin_tags    |
|sections.description |	:x:   |
|rating	|   meta.rating.value / rating  |
|reviews_link	|   :x:   |
|author	|   author / plugin.author  |
|author_profile	|   ~ author_login  |
|demo_url	|   :x:   |
|documentation_url	|   :x:   |
|version	|   tagged_versions?    |
|last_updated	|   plugin_modified |
|product_video	|   :x:   |
|is_featured	|   :x:   |
|setup_url	|   :x:   |

**Comment by gcsecsey** [comment]
@epeicher I've done some testing in Kibana, and I think these fields should also be available:

Field names returned by /marketplace/products | Field names returned by ES results (wp.org)
-- | --
id | post_id
sections.description | plugin.description

Other than these, I agree that we can't get the rest of the fields from ES for WP.org plugins yet.

**Comment by epeicher** [comment]
This is a similar table but includes an additional column with the `woocommerce` fields:

| Field names returned by `/marketplace/products` | Field names returned by ES results (wp.org) | wc.com fields |
| ----------------------------------------------- | ------------------------------------------- |------------------------------------------- |
| id                                              | post_id                                     |p.post_id                                     |
| name                                            | title.default                               |p.post_title                               |
| slug                                            | slug                                        |p.post_name                                        |
| org_slug                                        | :x:                                         |m.wpcom_marketplace_org_slug                                         |
| short_description                               | excerpt.en                                  |p.post_excerpt                                  |
| description                                     | description.en                              |:question:                          |
| requirements.plugins                            | :x:                                         |:question:                                         |
| requirements.required_primary_domain            | :x:                                         |:question:                                         |
| variations.monthly.product_id                   | :x:                                         |m.wpcom_marketplace_store_product_monthly_id                                         |
| variations.monthly.product_slug                 | :x:                                         |:question:                                         |
| variations.yearly.product_id                    | :x:                                         |m.wpcom_marketplace_store_product_yearly_id                                         |
| variations.yearly.product_slug                  | :x:                                         |:question:                                         |
| icons.1x                                        | plugin.icons                                |m._app_icon                                |
| icons.2x                                        | plugin.icons                                |m._app_icon                                |
| banners.high                                    | plugin.banners                              |m._heading_value                                            |
| banners.low                                     | plugin.banners                              |m._heading_value                            |
| tags                                            | taxonomy.plugin_tags                        |:question:                |
| sections.description                            | plugin.description                          |:question:  **Note:** same value than description                |
| rating                                          | meta.rating.value / rating                  | m._wc_average_rating               |
| reviews_link                                    | :x:                                         | m.wpcom_marketplace_reviews_link                |
| author                                          | author / plugin.author                      |m._product_author|
| author_profile                                  | ~ author_login                              |m._product_author_url                |
| demo_url                                        | :x:                                         |m._demo_url                |
| documentation_url                               | :x:                                         |m._documentation_url                |
| version                                         | tagged_versions?                            |m._product_version                |
| last_updated                                    | plugin_modified                             |p.post_modified?                |
| product_video                                   | :x:                                         |m.wpcom_marketplace_featured_video                |
| is_featured                                     | :x:                                         |:question:                |
| setup_url                                       | :x:                                         |m.wpcom_marketplace_setup_url                |

The SQL query can be executed in `woocommerce` database to obtain those fields 2d864-pb

**Comment by cpapazoglou** [comment]
_If you are searching fields by querying, remember that you will not find the key if it doesn't have value_

- `description` should be `post_content`
- `requirements.plugins` is `m._wc_general_product_dependency_plugin`
- `requirements.required_primary_domain` can be added as boolean in WC.com OR we may use `SEO` category to identify all plugins in this category as in need of a primary domain. We would need to change it here https://github.com/Automattic/wp-calypso/blob/6f82e4d82d7db395d877b37c12cda02fb6b6d010/client/my-sites/plugins/plugin-details-CTA/CTA-button.jsx#L78-L80
- `variations.monthly.product_slug` and `variations.yearly.product_slug` are not used, can be omitted
- `tags` are similar to `wpcom_marketplace_categories` taxonomy
- `is_featured` can be omitted, it's not used

**Comment by epeicher** [comment]
Thanks @cpapazoglou ! Adding your comments to the list, the new table would be 
> **Note** The symbol 🔲 means that fields should be indexed but they are not yet deployed to WC.com. Those fields must be tested once the changes are deployed to WC.com  tomorrow p1662556929352919/1662556538.686699-slack-C4E0DVDB2 

| Field names returned by `/marketplace/products` | Field names returned by ES results (wp.org) | wc.com fields                                     |
| ----------------------------------------------- | ------------------------------------------- | ------------------------------------------------- |
| id                                              | post_id                                     | p.post_id                                         |
| name                                            | title.default                               | p.post_title                                      |
| slug                                            | slug                                        | p.post_name                                       |
| org_slug                                        | :black_square_button:                       | m.wpcom_marketplace_org_slug                      |
| short_description                               | excerpt.en                                  | p.post_excerpt                                    |
| description                                     | description.en                              | p.post_content                                    |
| requirements.plugins                            | :black_square_button:                       | :m._wc_general_product_dependency_plugin          |
| requirements.required_primary_domain            | #67502                                      | :question:                                        |
| variations.monthly.product_id                   | :x:                                         | m.wpcom_marketplace_store_product_monthly_id      |
| variations.monthly.product_slug                 | `not used`                                  | :question:                                        |
| variations.yearly.product_id                    | :x:                                         | m.wpcom_marketplace_store_product_yearly_id       |
| variations.yearly.product_slug                  | `not used`                                  | :question:                                        |
| icons.1x                                        | plugin.icons                                | m._app_icon                                       |
| icons.2x                                        | plugin.icons                                | m._app_icon                                       |
| banners.high                                    | plugin.banners                              | m._heading_value                                  |
| banners.low                                     | plugin.banners                              | m._heading_value                                  |
| tags                                            | taxonomy.plugin_tags                        | m.wpcom_marketplace_categories                    |
| sections.description                            | plugin.description                          | :question:  **Note:** same value than description |
| rating                                          | meta.rating.value / rating                  | m._wc_average_rating                              |
| reviews_link                                    | :x:                                         | m.wpcom_marketplace_reviews_link                  |
| author                                          | author / plugin.author                      | m._product_author                                 |
| author_profile                                  | ~ author_login                              | m._product_author_url                             |
| demo_url                                        | :x:                                         | m._demo_url                                       |
| documentation_url                               | :x:                                         | m._documentation_url                              |
| version                                         | tagged_versions?                            | m._product_version                                |
| last_updated                                    | plugin_modified                             | p.post_modified?                                  |
| product_video                                   | :black_square_button:                       | m._featured_product_video                         |
| is_featured                                     | `not used`                                  | :question:                                        |
| setup_url                                       | :x:                                         | m.wpcom_marketplace_setup_url                     |

Which leaves the following fields to be required to be indexed:
- m.wpcom_marketplace_store_product_monthly_id
- m.wpcom_marketplace_store_product_yearly_id
- m.wpcom_marketplace_reviews_link
- m._demo_url
- m._documentation_url
- m.wpcom_marketplace_setup_url

**Comment by epeicher** [comment]
@robfelty , the following list of meta fields in WC.com is used in the product details page but it is not currently returned by the `jetpack-search` index. Should I create an issue in Jetpack adding the following list fields to the [`postmeta_to_sync` fields](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/modules/class-search.php#L70)?

- wpcom_marketplace_store_product_monthly_id
- wpcom_marketplace_store_product_yearly_id
- wpcom_marketplace_reviews_link
- _demo_url
- _documentation_url
- wpcom_marketplace_setup_url

cc @cpapazoglou

**Comment by robfelty** [comment]
In general, I think we want to keep the number of meta fields to sync relatively small, and used for common use cases. I wonder if in this case we could map these to the `jetpack-search-meta` fields instead. What do you think @gibrown ?

**Comment by cpapazoglou** [comment]
Would it make sense consolidating all fields in WooCommerce.com side to one field eg `wpcom_marketplace_fields`? This would help us 

- add only this field in jetpack sync fields
- enhancing `wpcom_marketplace_fields` to include more fields in the future if needed without having to change the indexed fields in Jetpack.

**Comment by robfelty** [comment]
So you want to add it as a PHP-serialized array? (or json blob)? I suppose that could work. It may not work on the indexing side though, depending on how the fields are defined there, since we don't allow introduction of dynamic fields in general, but there are some specific templates which allow some dynamic fields, like taxonomy, and `has.*`

**Comment by cpapazoglou** [comment]
> So you want to add it as a PHP-serialized array? (or json blob)? I suppose that could work. It may not work on the indexing side though, depending on how the fields are defined there, since we don't allow introduction of dynamic fields in general, but there are some specific templates which allow some dynamic fields, like taxonomy, and `has.*`

That's correct. These are not fields that would be used for search that's why I proposed it, jetpack index can skip them we just need them synced in the shadow site.

**Comment by robfelty** [comment]
okay - then putting them all into one field sounds like a good idea to me.

**Comment by gibrown** [comment]
FWIW, I think freely adding these fields is fine. The main goal is to not end up with more than 100k fields. It would be good if these fields were a bit more generic, but I don't think we should worry too much about them. As long as we don't auto add every field (like we used to) then I doubt we will run into a problem.

**Comment by epeicher** [comment]
I have created a task https://github.com/Automattic/wp-calypso/issues/67863 to investigate/implement a consolidated field with a serialized object containing Marketplace specific fields

**Comment by epeicher** [comment]
The [suggested approach](https://github.com/Automattic/wp-calypso/issues/64891#issuecomment-1243817089) to get the missing fields will be implemented as part of
- https://github.com/Automattic/wp-calypso/issues/68524

**Comment by epeicher** [comment]
I have created the following draft diff D91105-code to add the new serialised field to the index so it can be returned by the ES results

**Comment by epeicher** [comment]
Diff deployed r256039-wpcom

**Comment by cpapazoglou** [comment]
Duplicate of https://github.com/Automattic/dotcom-forge/issues/1482
